pkg/looker.go

```diff
@@ -0,0 +1,170 @@
+package pkg
+
+import (
+	"concierge/config"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/parnurzeal/gorequest"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+var client *LookerClient
+
+type LookerClient struct {
+	accessToken         string
+	accessTokenExpireAt int64
+	baseUrl             string
+	clientId            string
+	clientSecret        string
+}
+
+type LookerPatchUserRequest struct {
+	IsDisabled bool `json:"is_disabled"`
+}
+
+type LookerPatchUserResponse struct {
+	IsDisabled bool `json:"is_disabled"`
+}
+
+type LookerSearchUserRequest struct {
+	Email string `json:"email"`
+}
+
+type LookerSearchUserResponse struct {
+	Id         int  `json:"id"`
+	IsDisabled bool `json:"is_disabled"`
+}
+
+func GetLookerClient() *LookerClient {
+	if client == nil {
+		client = &LookerClient{
+			accessToken:         "",
+			accessTokenExpireAt: 0,
+			baseUrl:             config.LookerConfig.BaseUrl,
+			clientId:            config.LookerConfig.ClientId,
+			clientSecret:        config.LookerConfig.ClientSecret,
+		}
+	}
+	return client
+}
+
+func (c *LookerClient) PatchUser(userId int, req LookerPatchUserRequest) (*LookerPatchUserResponse, error) {
+	path := fmt.Sprintf("api/3.1/users/%d", userId)
+	method := http.MethodPatch
+	body := struct {
+		IsDisabled bool `json:"is_disabled"`
+	}{req.IsDisabled}
+
+	resp := LookerPatchUserResponse{}
+
+	httpResponse, httpErr := c.executeRequest(path, method, body)
+
+	if httpErr != nil {
+		return nil, httpErr
+	}
+
+	if decodeErr := json.NewDecoder(httpResponse.Body).Decode(&resp); decodeErr != nil {
+		return nil, decodeErr
+	}
+	return &resp, nil
+}
+
+func (c *LookerClient) SearchUser(req LookerSearchUserRequest) ([]LookerSearchUserResponse, error) {
+	var resp []LookerSearchUserResponse
+
+	path := fmt.Sprintf("api/3.1/users/search?email=%s", url.QueryEscape(req.Email))
+
+	httpResponse, httpErr := c.executeRequest(path, http.MethodGet, nil)
+
+	if httpErr != nil {
+		return nil, httpErr
+	}
+
+	if decodeErr := json.NewDecoder(httpResponse.Body).Decode(&resp); decodeErr != nil {
+		return nil, decodeErr
+	}
+
+	return resp, nil
+}
+
+func (c *LookerClient) isAccessTokenExpired() bool {
+	return time.Now().Unix() >= c.accessTokenExpireAt
+}
+
+func (c *LookerClient) setAccessToken() error {
+	requestBody := struct {
+		ClientId     string `json:"client_id"`
+		ClientSecret string `json:"client_secret"`
+	}{c.clientId, c.clientSecret}
+
+	requestJson, _ := json.Marshal(&requestBody)
+
+	requestString := string(requestJson)
+
+	httpResponse, _, errs := gorequest.New().
+		Post(c.baseUrl + "login").
+		Type(gorequest.TypeForm).
+		Send(requestString).
+		End()
+
+	if errs != nil {
+		return errs[0]
+	}
+
+	if httpResponse.StatusCode >= 400 {
+		return errors.New("failed to enable user on looker")
```

Comment:
more details on the error will definitely help

similarly in other places too

Reply:
right now we are calling 2 looker endpoints.

the only times it generates status code >400 are

1) malformed json
~~~json
{
    "message": "Problems parsing JSON",
    "documentation_url": "http://docs.looker.com/"
}
~~~

2) failed authentication
~~~json
{
    "message": "Not found",
    "documentation_url": "http://docs.looker.com/"
}
~~~

now malformed json is unlikely as we are using structs -> json directly and not doing construction by hand

failed authentication, the error message is not useful

dont want to  passthrough error message returned by looker as it can contain potentially contain sensitive fields. wasnt able to generate other errors in the endpoints we are calling
---------


will split the errors into 4xx and 5xx though and make it more descriptive.
at business logic level, the errors are more descriptive